### ISSUE TITLE
CFP: Mutual Authentication for Service Mesh

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#22215]
This meta issue tracks progress on the [Next-Generation Mutual Authentication with Cilium Service Mesh](https://isovalent.com/blog/post/2022-05-03-servicemesh-security/) implementation.

The [CFP](https://docs.google.com/document/d/1Do9_ICj4ZokSiWqXYdb7tQvhR--Rz2GyHJ2Xk6NDhQU/edit?usp=sharing) describes how we are planning to implement this, why we've chosen SPIFFE and SPIRE over something less complex, and so on. Here, we track actually getting the work done.

Tasks (this list will be updated as we add more tasks):
- [x] Discuss the [CFP](https://docs.google.com/document/d/1Do9_ICj4ZokSiWqXYdb7tQvhR--Rz2GyHJ2Xk6NDhQU/edit?usp=sharing) with the community
  - [x] Add to design-CFPs repo https://github.com/cilium/design-cfps/pull/5 
- [x] Implementation (pre code-freeze for 1.14)
  - [x] #23801
  - [x] #23802
  - [x] #23804
  - [x] #23805
  - [x] #23806
  - [x] #24897 
  - [x] #23807
  - [x] #25213
  - [x] #25925
  - [x] #25475
  - [x] #25227
  - [x] #24552
  - [x] #24600
  - [x] #24867
  - [x] #25860
  - [x] Feature integration testing
- [x] Post-code freeze for 1.14
  - [x] #24598
  - [x] #25216
  - [x] #25898
  - [x] #25899
- [x] Documentation
	- [x] Conceptual documentation (how the feature works)
	- [x] Getting started guide (how to start using the feature)
	- [x] Policy documentation (what we've added to Policy)
	- [x] Other documentation TBD
- [ ] Moved to 1.15
  - [x] #23808
  - [ ] #25730 (this actually requires a more nuanced discussion, but this feature will _not_ default to on for 1.14).
  - [x] #26480 
  - [ ] #26478
  - [x] #24617

What we are doing in this issue builds on the previous work done in #17335, but we are not _currently_ building out support for using a SPIFFE identity as a policy _target_, the identities used for mTLS will be largely invisible to the user for now. (We've definitely discussed doing this in the future though).

**Comment by abvaidya** [comment]
This is awesome. Would Cilium installation allow a choice to use a different SPIFFE implementation than SPIRE or would it install / require SPIRE?

**Comment by youngnick** [comment]
The initial release will include a way to install SPIRE for you, that we allow you to disable. But we probably won't have tutorial docs for bring-your-own-SPIRE for a bit. It is definitely something we want to allow though.

**Comment by abvaidya** [comment]
Oh, I was actually looking for using a different SPIFFE implementation than SPIRE. From what I read in CFP, Cilium will have a runtime dependency on SPIRE, is that correct? Is there any way I can use the new feature when it's available without SPIRE?

**Comment by youngnick** [comment]
Sorry @abvaidya, I missed your response. A lot of our code currently _assumes_ you're using SPIRE, but we're only using SPIFFE APIs, so theoretically you could swap this out for another SPIFFE implementation. I'd be interested to hear more though - perhaps you could log an issue with what you'd like?

**Comment by abvaidya** [comment]
no worries. This time it was my turn to miss the response. :)
Sure I would be happy to create an issue. I am a maintainer of Athenz - also a CNCF project in the service identity and authorization space. Athenz issued X.509 certificates contain SPIFFE URI and I was hoping to use this Cilium feature with Athenz fulfilling SPIFFE contract. So if its possible to swap out SPIRE it would be perfect.

**Comment by youngnick** [comment]
For those watching this issue: This feature is now code complete, and will be available in 1.14. It has test coverage, both at a unit and e2e/integration level, although it's not exhaustive, and we will have the usual set of documentation and tutorials available for it at launch.

I'd rate the overall support as at least Beta, shading towards Limited.

The items marked "post code freeze" will be completed before 1.14 final release, and the ones marked "deferred to 1.15" will be built during the 1.15 release.

**Comment by youngnick** [comment]
Okay everyone watching this, it seems that #26478 is waiting on some changes to how we handle BPF map pressure, so I'm moving it into the 1.15 timeframe.

That means that all the 1.14 deliverables for this are now done! 🎉

**Comment by youngnick** [comment]
#26478 has staled out, and it's moved to 1.15 anyway, so I'm going to close this one out. Don't worry, there are still heaps more improvements coming, but since this issue covers the initial delivery, it's done!
